File: backend/src/jenmoney/schemas/transfer.py
from datetime import datetime
from typing import Annotated, Any

from pydantic import Field, model_validator

from jenmoney.schemas.base import BaseAPIModel
from jenmoney.schemas.account import Currency


class TransferBase(BaseAPIModel):
    """Base transfer model with core fields."""

    from_account_id: int
    to_account_id: int
    from_amount: Annotated[float, Field(gt=0)]
    description: str | None = None

    @model_validator(mode="before")
    @classmethod
    def validate_different_accounts(cls, values: Any) -> Any:
        """Ensure transfer is between different accounts."""
        if isinstance(values, dict):
            from_account_id = values.get("from_account_id")
            to_account_id = values.get("to_account_id")

            if from_account_id == to_account_id:
                raise ValueError("Transfer must be between different accounts")

        return values


class TransferCreate(TransferBase):
    """Model for creating a new transfer."""

    # Optional: specify destination amount for multi-currency transfers
    to_amount: Annotated[float, Field(gt=0)] | None = None


class TransferUpdate(BaseAPIModel):
    """Model for updating a transfer."""

    description: str | None = None
    from_amount: Annotated[float, Field(gt=0)] | None = None
    to_amount: Annotated[float, Field(gt=0)] | None = None


class TransferResponse(TransferBase):
    """Model for transfer responses."""

    id: int
    to_amount: float
    from_currency: Currency
    to_currency: Currency
    exchange_rate: float | None = None
    created_at: datetime
    updated_at: datetime


class TransferListResponse(BaseAPIModel):
    """Paginated response for transfer listings."""

    items: list[TransferResponse]
    total: int
    page: int = 1
    size: int = 10
    pages: int

File: frontend/src/types/transfer.ts
import { Currency } from './account';

export interface Transfer {
  id: number;
  from_account_id: number;
  to_account_id: number;
  from_amount: number;
  to_amount: number;
  from_currency: Currency;
  to_currency: Currency;
  exchange_rate?: number | null;
  description?: string | null;
  created_at: string;
  updated_at: string;
}

export interface TransferCreate {
  from_account_id: number;
  to_account_id: number;
  from_amount: number;
  to_amount?: number; // Optional for auto-calculation in same currency
  description?: string | null;
}

export interface TransferFormData {
  from_account_id: number;
  to_account_id: number;
  from_amount: number;
  to_amount: number;
  description: string;
}

export interface TransferValidationErrors {
  from_account_id?: string;
  to_account_id?: string;
  from_amount?: string;
  to_amount?: string;
  description?: string;
  general?: string;
}